recipes/_windows.rb

```diff
@@ -16,98 +16,89 @@
-      checksum package.checksum
-      root_dir node['build-essential']['msys']['path']
+# The general layout is as such:
+# + mingw-prefix
+# |-  base   - holds common 32 and 64-bit msys packages.  Files from here are
+# |            copied into locations below.
+# |+  tool32 - 32-bit toolchain.
+#  |- .cache - holds temporary tar files and fetches/extracted packages.
+# |+  tool64 - 64-bit toolchain.
+#  |- .cache - holds temporary tar files and fetches/extracted packages.
+#
+# TODO: See if one can use /etc/fstab in base to mount tool32/tool64 as needed
+# instead of copying all the time.
+
+prefix = node['build-essential']['mingw']['prefix']
+base_path = ::File.join(prefix, 'base')
+tool32_path = ::File.join(prefix, 'tool32')
+tool64_path = ::File.join(prefix, 'tool64')
+
+potentially_at_compile_time do
+  directory 'mingw root directory' do
+    path prefix
+    action :create
+    recursive true
+  end
+
+  mingw_get 'msys core' do
+    package 'msys-base=2013072300-msys-bin.meta'
+    root base_path
+  end
+
+  mingw_get 'msys core extensions' do
+    package 'msys-core-utils-ext=5.97-3-*'
+    root base_path
+  end
+
+  mingw_get 'msys perl' do
+    package 'msys-perl-bin=5.8.8-*'
+    root base_path
+  end
+
+  directory tool32_path do
+  end
+
+  directory tool64_path do
+  end
+
+  ruby_block 'copy base files into tool32 and tool64' do
+    block do
+      ::FileUtils.cp_r("#{base_path}/.", tool32_path)
+      ::FileUtils.cp_r("#{base_path}/.", tool64_path)
     end
   end
+
+  mingw_tdm_gcc 'TDM GCC 32-bit with SJLJ' do
+    version '5.1.0'
+    flavor :sjlj_32
+    root tool32_path
+  end
+
+  mingw_tdm_gcc 'TDM GCC 64-bit with SJLJ/SEH' do
+    version '5.1.0'
+    flavor :seh_sjlj_64
+    root tool64_path
+  end
 end
 
-[
-  msys_p('http://iweb.dl.sourceforge.net/project/tdm-gcc/TDM-GCC%205%20series/5.1.0-tdm64-1/gcc-5.1.0-tdm64-1-core.tar.lzma',
-    '29393aac890847089ad1e93f81a28f6744b1609c00b25afca818f3903e42e4bd'),
-  msys_p('http://iweb.dl.sourceforge.net/project/tdm-gcc/MinGW-w64%20runtime/GCC%205%20series/mingw64runtime-v4-git20150618-gcc5-tdm64-1.tar.lzma',
-    '29186e0bb36824b10026d78bdcf238d631d8fc1d90718d2ebbd9ec239b6f94dd'),
-  msys_p('http://sourceforge.net/projects/tdm-gcc/files/GNU%20binutils/binutils-2.25-tdm64-1.tar.lzma',
-    '4722bb7b4d46cef714234109e25e5d1cfd29f4e53365b6d615c8a00735f60e40'),
-  msys_p('http://sourceforge.net/projects/tdm-gcc/files/TDM-GCC%205%20series/5.1.0-tdm64-1/gcc-5.1.0-tdm64-1-c++.tar.lzma',
-    '17fd497318d1ac187a113e8665330d746ad9607a0406ab2374db0d8e6f4094d1')
-].each do |package|
-  potentially_at_compile_time do
-    build_essential_msys_archive package.url do
-      root_dir node['build-essential']['msys']['path']
-      checksum package.checksum
-      mingw true
-    end
+# Clear out legacy compiler installation if present.
+# TODO: Remove this once all jenkins nodes have been converged with new changes
+# at least once.
+potentially_at_compile_time do
```

Comment:
I have mixed feelings on whether this belongs in this cookbook or some other wrapper cookbook. Thoughts @sersut?